File: Python.PassDetection/src/features/velocities_non_dominant_foot.py
from typing import List

from src.domain.common import Vector3
from src.domain.inferences import Target, BaseFeature
from src.domain.recordings import InputData


class VelocitiesNonDominantFoot(BaseFeature):
    @property
    def size(self) -> int:
        return 3  # Each feature now has only one component (x, y, or z)

    def calculate(self, input_data: InputData) -> List[Target]:
        non_dominant_positions = input_data.user_non_dominant_foot_positions
        timestamps = input_data.timestamps
        velocities = []

        for i in range(1, len(non_dominant_positions)):
            dt = timestamps[i] - timestamps[i - 1]
            if dt == 0:
                dt = 1e-6  # Avoid division by zero
            dx = (non_dominant_positions[i].x - non_dominant_positions[i - 1].x) / dt
            dy = (non_dominant_positions[i].y - non_dominant_positions[i - 1].y) / dt
            dz = (non_dominant_positions[i].z - non_dominant_positions[i - 1].z) / dt
            velocities.append(Vector3(dx, dy, dz))

        velocities.insert(0, Vector3(0, 0, 0))  # Insert a zero velocity at the start

        x_values = [vel.x for vel in velocities]
        y_values = [vel.y for vel in velocities]
        z_values = [vel.z for vel in velocities]

        return [
            Target(name="velocities_non_dominant_foot_x", values=x_values),
            Target(name="velocities_non_dominant_foot_y", values=y_values),
            Target(name="velocities_non_dominant_foot_z", values=z_values),
        ]

File: Unity.Interactions/Assets/_Project/PassDetection.Replay/Features/VelocitiesNonDominantFoot.cs
using System.Collections.Generic;
using System.Linq;
using _Project.PassDetection.Replay;
using _Project.PassDetection.Replay.Features;
using Tactive.MachineLearning._Project.MachineLearning;
using UnityEngine;

namespace Src.Domain.Inferences
{
    public class VelocitiesNonDominantFoot : BaseFeature<InputData>
    {
        public override int Size => 3;

        public override List<Feature> Calculate(InputData inputData)
        {
            var nonDominantPositions = inputData.UserNonDominantFootPositions;
            var timestamps = inputData.Timestamps;
            var velocities = new List<Vector3>();

            for (var i = 1; i < nonDominantPositions.Count; i++)
            {
                var dt = timestamps[i] - timestamps[i - 1];
                if (dt == 0) dt = 1e-6f;

                var dx = (nonDominantPositions[i].x - nonDominantPositions[i - 1].x) / dt;
                var dy = (nonDominantPositions[i].y - nonDominantPositions[i - 1].y) / dt;
                var dz = (nonDominantPositions[i].z - nonDominantPositions[i - 1].z) / dt;
                velocities.Add(new Vector3(dx, dy, dz));
            }

            velocities.Insert(0, Vector3.zero);

            var xValues = velocities.Select(v => v.x).ToList();
            var yValues = velocities.Select(v => v.y).ToList();
            var zValues = velocities.Select(v => v.z).ToList();

            return new List<Feature>
            {
                new Feature("velocities_non_dominant_foot_x", xValues),
                new Feature("velocities_non_dominant_foot_y", yValues),
                new Feature("velocities_non_dominant_foot_z", zValues)
            };
        }
    }
}

namespace Src.Domain.Inferences
{
}
